src/agent/Dispatcher.ts

```diff
@@ -40,9 +40,15 @@ class Dispatcher {
         outboundMessage.payload.setReturnRouting(ReturnRouteTypes.all)
       }
 
-      // check for return routing, with thread id
+      // Check for return routing, with thread id
       if (message.hasReturnRouting(threadId)) {
-        return await this.messageSender.packMessage(outboundMessage)
+        // Find service with highest priority and keys to pack message
+        const [service] = this.transportService.findServices(outboundMessage.connection)
+        if (!service) {
+          throw new AriesFrameworkError(`Connection with id ${outboundMessage.connection.id} has no service!`)
+        }
+        // TODO Should we validate inbound transport session scheme and service endpoint scheme?
+        return await this.messageSender.packMessage(outboundMessage, service)
```

Comment:
This doesn't make sense to me. Why are we using the first service entry for return routing? We should use the same keys that were used for receiving the message.

Reply:
Maybe packMessage shouldn't take a service as input, as you can pack messages without a service. You just need the keys

Reply:
Yes, I'm gradually refactoring this method. Now it's just about encryption of payload with keys. I just kept interface/signature of the method until I replace `createOutboundMessage` with OutboundMessageContext and update interface of `Agent` class.

Reply:
Usage of service has been removed.